### PR TITLE
[TorchDISC]supports dynamic shape in python unit test

### LINKED ISSUE #383: [TorchDISC] Enhancement torch-mlir converter UT to support configure shape info
Need to configure shape info to check converter is correct, including static shape, full dynamic shape, and partial dynamic shape.

### PR DESCRIPTION
**Pull request by Yancey1989**
fixed #383

### CODE REVIEW

File: pytorch_blade/tests/disc/ops/test_activation.py [hunk 1]
```diff
@@ -9,29 +9,50 @@
...
+        [ShapeTesting.FULL_DYNAMIC_SHAPE.name],
+        [ShapeTesting.STATIC_SHAPE.name],
+        [ShapeTesting.PARTIAL_DYNAMIC_SHAPE.name],
+    ])
+    def test_relu(self, name):
+        self._test_activation(torch.nn.ReLU(), torch.nn.functional.relu, name)
 
-        self._test_activation(relu_func)
 
+    @skipIfNoTorchMlir()
```
Comment (fortianyou) [inline pytorch_blade/tests/disc/ops/test_activation.py:55]: `skipIfNoTorchMlir` -> `skipIfIsTorchMlir`?

**Review by fortianyou — CHANGES_REQUESTED** [review]

File: pytorch_blade/src/compiler/mlir/converters/mhlo_conversion.cpp [hunk 2]
```diff
@@ -336,6 +336,8 @@ ConvertTorchToMhlo(std::shared_ptr<torch::jit::Graph> graph) {
       ::mlir::ModuleOp::create(::mlir::UnknownLoc::get(&mlir_context));
   torch::jit::CompilationUnit unit;
   auto fn = unit.create_function("main", graph);
+  // removeInputShape(graph);
+  std::cout << graph->toString() << std::endl;
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply pytorch_blade/src/compiler/mlir/converters/mhlo_conversion.cpp]: done, remove the unused log.

File: pytorch_blade/src/compiler/jit/pybind_functions.cpp [hunk 3]
```diff
@@ -66,6 +66,7 @@ void initToolsBindings(py::module& m) {
       "create_method_from_graph", &torch::blade::create_method_from_graph);
   tools.def("unsafe_remove_method", &torch::blade::unsafe_remove_method);
   tools.def("get_list_tensor_type", &torch::blade::get_list_tensor_type);
+  tools.def("tensor_to_dynamic_shape", torch::blade::tensor_to_dynamic_shape);
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply pytorch_blade/src/compiler/jit/pybind_functions.cpp]: done

File: pytorch_blade/tests/disc/testing_base.py [hunk 4]
```diff
@@ -42,26 +42,27 @@ def _ScriptFunction2Module(self, nn_module):
         else:
             return nn_module
 
-    def cvt_to_disc(self, nn_module, test_data):
+    def cvt_to_disc(self, nn_module, test_data, input_dims):
         cfg = Config.get_current_context_or_new()
         cfg.optimization_pipeline = mlir.backend_name()
+        cfg.force_input_dims = input_dims
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply pytorch_blade/tests/disc/testing_base.py]: I just renamed the attr-name, torch-mlir using decorator to config the input annotate, torch_blade using context guard as the configuration, we can do some refactor in the later if we want to enable decorator to do some configuration.

File: pytorch_blade/tests/disc/ops/test_activation.py (hunk 1 above)
Comment (Yancey1989) [reply pytorch_blade/tests/disc/ops/test_activation.py]: done.

**Comment by Yancey1989** [comment]
> Another concern about removing shape from exported graph: DISC currently DO remove shape info from input MHLO IR. Is this redundant to that? Or this is to make passes before DISC (in torch_blade) not depend on concrete shape?

@qiuxiafei not redundant, we also need to check the dynamic shape feature is valid on path lowering from torchscript to mhlo.

File: pytorch_blade/src/compiler/jit/fusion.cpp [hunk 5]
```diff
@@ -136,5 +136,20 @@ torch::TypePtr create_tensor_type_from_scalar_type(const c10::Type& typ) {
   return nullptr;
 }
 
+void tensor_to_dynamic_shape(
+    torch::jit::Value* val,
+    std::vector<int64_t>& dims) {
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply pytorch_blade/src/compiler/jit/fusion.cpp]: done.

File: pytorch_blade/torch_blade/config.py [hunk 6]
```diff
@@ -455,3 +456,11 @@ def my_custom_fuse_jit_pass(graph):
     def customize_jit_passes(self, val):
         assert isinstance(val, list), "customize_jit_passes should be list, got {}".format(type(val))
         self._customize_jit_passes = val
+    
+    @property
+    def force_input_dims(self):
+        return self._force_input_dims
+    
+    @force_input_dims.setter
+    def force_input_dims(self, dims):
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply pytorch_blade/torch_blade/config.py]: I added some input type checks in the setting function, so maybe we can keep the getting and setter function.

File: pytorch_blade/tests/disc/testing_base.py [hunk 7]
```diff
@@ -42,27 +44,41 @@ def _ScriptFunction2Module(self, nn_module):
...
         self.assertEqual(output, result, rtol=rtol, atol=atol)
         self.assertGreaterEqual(mlir.num_engines(opt_module), n_engines)
         return output, result
+
+    def _gen_test_data(self, dims):
+        test_data = []
+        for dim in dims:
+            for i, dim_i in enumerate(dim):
+                if dim_i == -1:
+                    dim[i] = random.randint(1, 100)
```
Comment (wyzero) [inline pytorch_blade/tests/disc/testing_base.py:78]: Should the seed the fixed, so that we can re-produce tests?

Comment (Yancey1989) [reply pytorch_blade/tests/disc/testing_base.py:79]: done.

**Review by fortianyou — APPROVED** [review]
LGTM
